File: tools/export_embedding.py
#!/usr/bin/env python3
"""
export_embedding.py — AnnData (scanpy)  ->  assets/data/embedding.json

Emits ONLY obsm coordinates and the obs columns you name. The counts matrix,
barcodes, and every unnamed obs column stay behind.

Usage
    python3 tools/export_embedding.py \
        --input      adata.h5ad \
        --output     assets/data/embedding.json \
        --reductions X_pca,X_umap \
        --labels     guide,phase \
        --scores     het_cv \
        --n          6000 \
        --source     "MDA-MB-231 CRISPRi Perturb-seq" \
        --caption    "6,000 cells from ..."

Only --input is required.

Requires anndata (and numpy/pandas, which come with it):
    pip install anndata
"""

from __future__ import annotations

import argparse
import json
import os
import sys

# Common obsm keys mapped to the short names the site uses in its UI.
PRETTY = {"X_pca": "pca", "X_umap": "umap", "X_tsne": "tsne", "X_diffmap": "diffmap"}


def parse_args(argv=None):
    p = argparse.ArgumentParser(
        description="Export an AnnData embedding to the site's embedding.json format.",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )
    p.add_argument("--input", required=True, help="path to a .h5ad file")
    p.add_argument("--output", default="assets/data/embedding.json")
    p.add_argument("--reductions", default="X_pca,X_umap",
                   help="comma-separated obsm keys; order matters, the last is the default view")
    p.add_argument("--cloud", default="",
                   help="obsm key for the hero's 3D intro cloud (dims 1-3); "
                        "defaults to the first --reductions entry")
    p.add_argument("--labels", default="", help="comma-separated categorical obs columns")
    p.add_argument("--scores", default="", help="comma-separated continuous obs columns")
    p.add_argument("--n", type=int, default=6000, help="max cells to export")
    p.add_argument("--seed", type=int, default=42, help="subsampling seed")
    p.add_argument("--source", default="", help="short description of the dataset")
    p.add_argument("--caption", default="", help="figure legend; auto-generated if omitted")
    p.add_argument("--digits", type=int, default=3, help="coordinate rounding")
    return p.parse_args(argv)


def split_csv(s: str) -> list[str]:
    return [t.strip() for t in s.split(",") if t.strip()]


def main(argv=None) -> int:
    args = parse_args(argv)

    try:
        import anndata as ad
        import numpy as np
        import pandas as pd
    except ImportError as e:
        sys.exit(f"missing dependency: {e.name}. Try:  pip install anndata")

    print(f"Reading {args.input} ...", file=sys.stderr)
    adata = ad.read_h5ad(args.input)

    reductions = split_csv(args.reductions)
    label_cols = split_csv(args.labels)
    score_cols = split_csv(args.scores)

    missing = [r for r in reductions if r not in adata.obsm]
    if missing:
        sys.exit(f"obsm key(s) not found: {', '.join(missing)}\n"
                 f"  available: {', '.join(adata.obsm.keys())}")

    missing = [c for c in label_cols + score_cols if c not in adata.obs.columns]
    if missing:
        sys.exit(f"obs column(s) not found: {', '.join(missing)}\n"
                 f"  available: {', '.join(adata.obs.columns)}")

    # --- deterministic subsample --------------------------------------------
    n_total = adata.n_obs
    rng = np.random.default_rng(args.seed)
    if n_total > args.n:
        keep = np.sort(rng.choice(n_total, size=args.n, replace=False))
    else:
        keep = np.arange(n_total)
    n = int(keep.size)
    print(f"Exporting {n} of {n_total} cells (seed {args.seed}).", file=sys.stderr)

    # --- 3D cloud for the hero intro ----------------------------------------
    cloud_key = args.cloud or reductions[0]
    coords3d = None
    if cloud_key in adata.obsm:
        m = np.asarray(adata.obsm[cloud_key])
        if m.shape[1] >= 3:
            coords3d = np.round(m[keep, :3].astype(float), args.digits).ravel().tolist()
            print(f"3D hero cloud from '{cloud_key}' dims 1-3.", file=sys.stderr)
        else:
            print(f"NOTE: '{cloud_key}' has <3 dimensions; hero will start flat.",
                  file=sys.stderr)

    # --- coordinates ---------------------------------------------------------
    coords = {}
    for key in reductions:
        m = np.asarray(adata.obsm[key])
        if m.shape[1] < 2:
            sys.exit(f"obsm['{key}'] has fewer than 2 dimensions")
        flat = np.round(m[keep, :2].astype(float), args.digits).ravel()
        coords[PRETTY.get(key, key.lstrip("X_"))] = flat.tolist()

    short = [PRETTY.get(k, k.lstrip("X_")) for k in reductions]

    # --- categorical label sets ---------------------------------------------
    label_sets = {}
    for col in label_cols:
        s = adata.obs[col].iloc[keep]
        cat = s.astype("category").cat.remove_unused_categories()
        levels = [str(x) for x in cat.cat.categories]
        if len(levels) > 24:
            print(f"WARNING: '{col}' has {len(levels)} levels; the legend will be "
                  f"unwieldy. Consider collapsing it.", file=sys.stderr)
        label_sets[col] = {
            "label": col,
            "levels": levels,
            "values": [int(v) for v in cat.cat.codes],  # already zero-indexed
        }

    # --- continuous scores ---------------------------------------------------
    scores = {}
    for col in score_cols:
        v = pd.to_numeric(adata.obs[col].iloc[keep], errors="coerce").to_numpy(dtype=float)
        if np.isnan(v).any():
            print(f"WARNING: '{col}' contains NaN; replacing with the column median.",
                  file=sys.stderr)
            v = np.nan_to_num(v, nan=float(np.nanmedian(v)))
        v = np.round(v, args.digits)
        scores[col] = {
            "label": col,
            "range": [float(v.min()), float(v.max())],
            "values": v.tolist(),
        }

    # --- assemble ------------------------------------------------------------
    caption = args.caption or (
        f"{n:,} cells. {short[-1].upper()} embedding; colour by "
        f"{label_cols[0] if label_cols else 'cluster'}."
    )

    payload = {
        "meta": {
            "n": n,
            "source": args.source,
            "caption": caption,
            "heroCaption": f"{n:,} cells · {short[-1].upper()}",
            "embeddingLabels": {s: s.upper() for s in short},
            "exportedBy": "tools/export_embedding.py",
            "seed": args.seed,
        },
        "coords": coords,
        "coords3d": coords3d,
        "labelSets": label_sets,
        "scores": scores,
    }

    os.makedirs(os.path.dirname(os.path.abspath(args.output)), exist_ok=True)
    with open(args.output, "w") as fh:
        json.dump(payload, fh, separators=(",", ":"))

    kb = round(os.path.getsize(args.output) / 1024)
    print(f"Wrote {args.output} ({kb} KB, {n} cells).", file=sys.stderr)
    if kb > 1024:
        print("NOTE: over 1 MB. Lower --n so the page still paints quickly.", file=sys.stderr)
    print(f"Validate with:  python3 tools/validate_embedding.py {args.output}", file=sys.stderr)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/export_embedding.R
#!/usr/bin/env Rscript
# =============================================================================
# export_embedding.R — Seurat object  ->  assets/data/embedding.json
# -----------------------------------------------------------------------------
# Emits ONLY reduction coordinates and the metadata columns you name.
# The counts matrix, barcodes, and every unnamed metadata column stay behind.
#
# Usage
#   Rscript tools/export_embedding.R \
#     --input      seurat.rds \
#     --output     assets/data/embedding.json \
#     --reductions pca,umap \
#     --labels     guide,Phase \
#     --scores     het_cv \
#     --n          6000 \
#     --source     "MDA-MB-231 CRISPRi Perturb-seq" \
#     --caption    "6,000 cells from ..."
#
# Only --input is required. Defaults are listed in the parser below.
# =============================================================================

suppressPackageStartupMessages({
  ok <- requireNamespace("jsonlite", quietly = TRUE)
})
if (!ok) stop("jsonlite is required:  install.packages('jsonlite')", call. = FALSE)

# --- argument parsing --------------------------------------------------------

parse_args <- function(argv) {
  defaults <- list(
    input      = NULL,
    output     = "assets/data/embedding.json",
    reductions = "pca,umap",
    cloud      = "",
    labels     = "",
    scores     = "",
    n          = "6000",
    seed       = "42",
    source     = "",
    caption    = "",
    digits     = "3"
  )
  i <- 1
  while (i <= length(argv)) {
    key <- sub("^--", "", argv[[i]])
    if (!key %in% names(defaults)) stop("unknown argument: ", argv[[i]], call. = FALSE)
    if (i + 1 > length(argv)) stop("missing value for --", key, call. = FALSE)
    defaults[[key]] <- argv[[i + 1]]
    i <- i + 2
  }
  if (is.null(defaults$input)) {
    stop("--input <seurat.rds> is required", call. = FALSE)
  }
  defaults$n      <- as.integer(defaults$n)
  defaults$seed   <- as.integer(defaults$seed)
  defaults$digits <- as.integer(defaults$digits)
  defaults
}

split_csv <- function(x) {
  if (!nzchar(x)) return(character(0))
  trimws(strsplit(x, ",", fixed = TRUE)[[1]])
}

args <- parse_args(commandArgs(trailingOnly = TRUE))

# --- load --------------------------------------------------------------------

message("Reading ", args$input, " ...")
obj <- readRDS(args$input)

if (!inherits(obj, "Seurat")) {
  stop("--input must be a saved Seurat object (got class: ",
       paste(class(obj), collapse = "/"), ")", call. = FALSE)
}

reductions <- split_csv(args$reductions)
# The hero's 3D intro cloud: first three dims of this reduction. Defaults to
# the first listed reduction (normally PCA, where PC3 is meaningful).
cloud_red <- if (nzchar(args$cloud)) args$cloud else reductions[1]
label_cols <- split_csv(args$labels)
score_cols <- split_csv(args$scores)

available <- names(obj@reductions)
missing_red <- setdiff(reductions, available)
if (length(missing_red)) {
  stop("reduction(s) not found: ", paste(missing_red, collapse = ", "),
       "\n  available: ", paste(available, collapse = ", "), call. = FALSE)
}

meta <- obj@meta.data
missing_meta <- setdiff(c(label_cols, score_cols), colnames(meta))
if (length(missing_meta)) {
  stop("metadata column(s) not found: ", paste(missing_meta, collapse = ", "),
       "\n  available: ", paste(colnames(meta), collapse = ", "), call. = FALSE)
}

# --- deterministic subsample -------------------------------------------------

n_total <- ncol(obj)
set.seed(args$seed)
keep <- if (n_total > args$n) sort(sample.int(n_total, args$n)) else seq_len(n_total)
n <- length(keep)
message("Exporting ", n, " of ", n_total, " cells (seed ", args$seed, ").")

# --- coordinates -------------------------------------------------------------

round_flat <- function(m, digits) {
  # Interleave to [x0,y0,x1,y1,...] and round.
  as.numeric(round(t(m[, 1:2, drop = FALSE]), digits))
}

# 3D cloud for the hero intro — skipped if the reduction has <3 dimensions.
coords3d <- NULL
if (cloud_red %in% available) {
  cloud_emb <- obj@reductions[[cloud_red]]@cell.embeddings
  if (ncol(cloud_emb) >= 3) {
    coords3d <- as.numeric(round(t(cloud_emb[keep, 1:3, drop = FALSE]), args$digits))
    message("3D hero cloud from '", cloud_red, "' dims 1-3.")
  } else {
    message("NOTE: '", cloud_red, "' has <3 dimensions; hero will start flat.")
  }
}

coords <- list()
for (r in reductions) {
  emb <- obj@reductions[[r]]@cell.embeddings
  if (ncol(emb) < 2) stop("reduction '", r, "' has fewer than 2 dimensions", call. = FALSE)
  coords[[r]] <- round_flat(emb[keep, , drop = FALSE], args$digits)
}

# --- label sets (categorical) ------------------------------------------------

label_sets <- list()
for (col in label_cols) {
  v <- meta[[col]][keep]
  f <- if (is.factor(v)) droplevels(v) else factor(v)
  lv <- levels(f)
  if (length(lv) > 24) {
    warning("column '", col, "' has ", length(lv),
            " levels; the legend will be unwieldy. Consider collapsing it.")
  }
  label_sets[[col]] <- list(
    label  = col,
    levels = as.character(lv),
    values = as.integer(f) - 1L            # zero-indexed for JS
  )
}

# --- scores (continuous) -----------------------------------------------------

scores <- list()
for (col in score_cols) {
  v <- as.numeric(meta[[col]][keep])
  if (anyNA(v)) {
    warning("column '", col, "' contains NA; replacing with the column median.")
    v[is.na(v)] <- stats::median(v, na.rm = TRUE)
  }
  scores[[col]] <- list(
    label  = col,
    range  = round(as.numeric(range(v)), args$digits),
    values = round(v, args$digits)
  )
}

# --- assemble ----------------------------------------------------------------

caption <- if (nzchar(args$caption)) args$caption else {
  paste0(n, " cells. ", toupper(reductions[length(reductions)]),
         " embedding; color by ",
         if (length(label_cols)) label_cols[1] else "cluster", ".")
}

payload <- list(
  meta = list(
    n               = n,
    source          = args$source,
    caption         = caption,
    heroCaption     = paste0(format(n, big.mark = ","), " cells · ",
                             toupper(reductions[length(reductions)])),
    embeddingLabels = setNames(as.list(toupper(reductions)), reductions),
    exportedBy      = "tools/export_embedding.R",
    seed            = args$seed
  ),
  coords    = coords,
  coords3d  = coords3d,
  labelSets = label_sets,
  scores    = scores
)

dir.create(dirname(args$output), recursive = TRUE, showWarnings = FALSE)
jsonlite::write_json(payload, args$output, auto_unbox = TRUE, digits = args$digits)

size_kb <- round(file.info(args$output)$size / 1024)
message("Wrote ", args$output, " (", size_kb, " KB, ", n, " cells).")
if (size_kb > 1024) {
  message("NOTE: over 1 MB. Lower --n so the page still paints quickly.")
}
message("Validate with:  python3 tools/validate_embedding.py ", args$output)
